File: laravel-backend/app/Services/TicketService.php
<?php

namespace App\Services;

use Illuminate\Support\Facades\DB;
use Illuminate\Support\Str;

class TicketService
{
    /**
     * Create a new ticket batch
     */
    public function createBatch(array $data): array
    {
        // Verify revenue point exists
        $revenuePoint = DB::connection('tenant')
            ->table('revenue_points')
            ->where('id', $data['revenue_point_id'])
            ->where('is_active', true)
            ->first();

        if (!$revenuePoint) {
            return ['success' => false, 'message' => 'Revenue point not found or inactive'];
        }

        // Verify revenue item exists
        $revenueItem = DB::connection('tenant')
            ->table('revenue_items')
            ->where('id', $data['revenue_item_id'])
            ->where('is_active', true)
            ->first();

        if (!$revenueItem) {
            return ['success' => false, 'message' => 'Revenue item not found or inactive'];
        }

        // Generate batch code
        $batchCode = $this->generateBatchCode($revenuePoint->code);

        // Calculate totals
        $totalTickets = $data['end_number'] - $data['start_number'] + 1;
        $expectedAmount = $totalTickets * $data['unit_price'];

        $batchId = DB::connection('tenant')->table('ticket_batches')->insertGetId([
            'batch_code' => $batchCode,
            'revenue_point_id' => $data['revenue_point_id'],
            'revenue_item_id' => $data['revenue_item_id'],
            'assigned_to' => $data['assigned_to'] ?? null,
            'assigned_to_type' => $data['assigned_to_type'] ?? 'user',
            'start_number' => $data['start_number'],
            'end_number' => $data['end_number'],
            'total_tickets' => $totalTickets,
            'unit_price' => $data['unit_price'],
            'expected_amount' => $expectedAmount,
            'valid_from' => $data['valid_from'],
            'valid_to' => $data['valid_to'],
            'status' => 'active',
            'created_by' => $data['created_by'] ?? null,
            'created_at' => now(),
            'updated_at' => now(),
        ]);

        // Create individual tickets
        $this->createTicketsForBatch($batchId, $data['start_number'], $data['end_number'], $data['unit_price']);

        $this->logAudit('create', 'tickets', 'TicketBatch', $batchId, [
            'batch_code' => $batchCode,
            'revenue_point' => $revenuePoint->name,
            'total_tickets' => $totalTickets,
            'expected_amount' => $expectedAmount,
        ]);

        return [
            'success' => true,
            'batch_id' => $batchId,
            'batch_code' => $batchCode,
            'total_tickets' => $totalTickets,
            'expected_amount' => $expectedAmount,
        ];
    }

    /**
     * Create individual tickets for a batch
     */
    protected function createTicketsForBatch(int $batchId, int $startNumber, int $endNumber, float $unitPrice): void
    {
        $batch = DB::connection('tenant')
            ->table('ticket_batches')
            ->where('id', $batchId)
            ->first();

        $tickets = [];
        $prefix = $this->getSetting('ticket_prefix', 'TKT');

        for ($i = $startNumber; $i <= $endNumber; $i++) {
            $ticketNumber = sprintf('%s-%s-%06d', $prefix, $batch->batch_code, $i);
            $qrCode = $this->generateQRCode($ticketNumber);

            $tickets[] = [
                'batch_id' => $batchId,
                'ticket_number' => $ticketNumber,
                'qr_code' => $qrCode,
                'amount' => $unitPrice,
                'status' => 'available',
                'synced' => true,
                'created_at' => now(),
                'updated_at' => now(),
            ];

            // Insert in chunks
            if (count($tickets) >= 100) {
                DB::connection('tenant')->table('tickets')->insert($tickets);
                $tickets = [];
            }
        }

        // Insert remaining
        if (!empty($tickets)) {
            DB::connection('tenant')->table('tickets')->insert($tickets);
        }
    }

    /**
     * Sell/Issue a ticket
     */
    public function sellTicket(array $data): array
    {
        $ticket = DB::connection('tenant')
            ->table('tickets')
            ->where('id', $data['ticket_id'])
            ->first();

        if (!$ticket) {
            return ['success' => false, 'message' => 'Ticket not found'];
        }

        if ($ticket->status !== 'available') {
            return ['success' => false, 'message' => 'Ticket is not available for sale'];
        }

        // Verify batch is valid
        $batch = DB::connection('tenant')
            ->table('ticket_batches')
            ->where('id', $ticket->batch_id)
            ->first();

        if (!$batch || $batch->status !== 'active') {
            return ['success' => false, 'message' => 'Batch is not active'];
        }

        $now = now();
        if ($now < $batch->valid_from || $now > $batch->valid_to) {
            return ['success' => false, 'message' => 'Ticket is outside validity period'];
        }

        // Verify seller is assigned to this batch
        $userId = $data['sold_by'] ?? auth()->id();
        if ($batch->assigned_to && $batch->assigned_to != $userId) {
            return ['success' => false, 'message' => 'You are not assigned to this batch'];
        }

        // Update ticket
        DB::connection('tenant')->table('tickets')
            ->where('id', $ticket->id)
            ->update([
                'status' => 'sold',
                'payment_method' => $data['payment_method'] ?? 'cash',
                'payment_reference' => $data['payment_reference'] ?? null,
                'sold_by' => $userId,
                'sold_at' => now(),
                'payer_name' => $data['payer_name'] ?? null,
                'payer_phone' => $data['payer_phone'] ?? null,
                'vehicle_number' => $data['vehicle_number'] ?? null,
                'synced' => $data['synced'] ?? true,
                'synced_at' => ($data['synced'] ?? true) ? now() : null,
                'updated_at' => now(),
            ]);

        // Update batch counters
        DB::connection('tenant')->table('ticket_batches')
            ->where('id', $batch->id)
            ->increment('tickets_sold');

        DB::connection('tenant')->table('ticket_batches')
            ->where('id', $batch->id)
            ->increment('amount_collected', $ticket->amount);

        // Create payment record
        DB::connection('tenant')->table('ticket_payments')->insert([
            'ticket_id' => $ticket->id,
            'amount' => $ticket->amount,
            'payment_method' => $data['payment_method'] ?? 'cash',
            'payment_reference' => $data['payment_reference'] ?? null,
            'status' => 'success',
            'collected_by' => $userId,
            'created_at' => now(),
            'updated_at' => now(),
        ]);

        $this->logAudit('sell', 'tickets', 'Ticket', $ticket->id, [
            'ticket_number' => $ticket->ticket_number,
            'amount' => $ticket->amount,
            'payment_method' => $data['payment_method'] ?? 'cash',
        ]);

        return [
            'success' => true,
            'ticket_number' => $ticket->ticket_number,
            'qr_code' => $ticket->qr_code,
            'amount' => $ticket->amount,
        ];
    }

    /**
     * Sell next available ticket from a batch
     */
    public function sellNextTicket(int $batchId, array $data): array
    {
        // Get next available ticket
        $ticket = DB::connection('tenant')
            ->table('tickets')
            ->where('batch_id', $batchId)
            ->where('status', 'available')
            ->orderBy('id', 'asc')
            ->first();

        if (!$ticket) {
            return ['success' => false, 'message' => 'No available tickets in this batch'];
        }

        return $this->sellTicket(array_merge($data, ['ticket_id' => $ticket->id]));
    }

    /**
     * Verify a ticket
     */
    public function verifyTicket(string $ticketNumber): array
    {
        $ticket = DB::connection('tenant')
            ->table('tickets')
            ->join('ticket_batches', 'tickets.batch_id', '=', 'ticket_batches.id')
            ->join('revenue_points', 'ticket_batches.revenue_point_id', '=', 'revenue_points.id')
            ->where('tickets.ticket_number', $ticketNumber)
            ->orWhere('tickets.qr_code', $ticketNumber)
            ->select('tickets.*', 'ticket_batches.valid_from', 'ticket_batches.valid_to', 
                     'revenue_points.name as revenue_point_name')
            ->first();

        if (!$ticket) {
            return [
                'success' => false,
                'valid' => false,
                'message' => 'Ticket not found',
            ];
        }

        $now = now();
        $isValid = $ticket->status === 'sold' 
                   && $now >= $ticket->valid_from 
                   && $now <= $ticket->valid_to;

        // Mark as verified
        if ($isValid && $ticket->status === 'sold') {
            DB::connection('tenant')->table('tickets')
                ->where('id', $ticket->id)
                ->update([
                    'status' => 'verified',
                    'verified_at' => now(),
                    'verified_by' => auth()->id(),
                    'updated_at' => now(),
                ]);
        }

        return [
            'success' => true,
            'valid' => $isValid,
            'ticket' => [
                'ticket_number' => $ticket->ticket_number,
                'status' => $ticket->status,
                'amount' => $ticket->amount,
                'sold_at' => $ticket->sold_at,
                'valid_from' => $ticket->valid_from,
                'valid_to' => $ticket->valid_to,
                'revenue_point' => $ticket->revenue_point_name,
                'payer_name' => $ticket->payer_name,
                'vehicle_number' => $ticket->vehicle_number,
            ],
            'message' => $isValid ? 'Ticket is valid' : 'Ticket is invalid or expired',
        ];
    }

    /**
     * Request ticket cancellation
     */
    public function requestCancellation(int $ticketId, string $reason): array
    {
        $ticket = DB::connection('tenant')
            ->table('tickets')
            ->where('id', $ticketId)
            ->first();

        if (!$ticket) {
            return ['success' => false, 'message' => 'Ticket not found'];
        }

        if (!in_array($ticket->status, ['sold', 'available'])) {
            return ['success' => false, 'message' => 'Ticket cannot be cancelled'];
        }

        DB::connection('tenant')->table('tickets')
            ->where('id', $ticketId)
            ->update([
                'status' => 'cancelled',
                'cancellation_reason' => $reason,
                'cancelled_by' => auth()->id(),
                'cancelled_at' => now(),
                'updated_at' => now(),
            ]);

        // Update batch counters
        $batch = DB::connection('tenant')
            ->table('ticket_batches')
            ->where('id', $ticket->batch_id)
            ->first();

        DB::connection('tenant')->table('ticket_batches')
            ->where('id', $batch->id)
            ->increment('tickets_cancelled');

        if ($ticket->status === 'sold') {
            DB::connection('tenant')->table('ticket_batches')
                ->where('id', $batch->id)
                ->decrement('tickets_sold');

            DB::connection('tenant')->table('ticket_batches')
                ->where('id', $batch->id)
                ->decrement('amount_collected', $ticket->amount);
        }

        $this->logAudit('cancel', 'tickets', 'Ticket', $ticketId, [
            'ticket_number' => $ticket->ticket_number,
            'reason' => $reason,
        ]);

        return ['success' => true, 'message' => 'Ticket cancelled successfully'];
    }

    /**
     * Sync offline tickets
     */
    public function syncOfflineTickets(array $tickets): array
    {
        $results = [];

        foreach ($tickets as $ticketData) {
            $ticket = DB::connection('tenant')
                ->table('tickets')
                ->where('id', $ticketData['ticket_id'])
                ->first();

            if (!$ticket) {
                $results[$ticketData['ticket_id']] = ['success' => false, 'message' => 'Ticket not found'];
                continue;
            }

            // Update ticket with offline data
            DB::connection('tenant')->table('tickets')
                ->where('id', $ticket->id)
                ->update([
                    'status' => 'sold',
                    'payment_method' => $ticketData['payment_method'] ?? 'cash',
                    'sold_by' => $ticketData['sold_by'],
                    'sold_at' => $ticketData['sold_at'],
                    'payer_name' => $ticketData['payer_name'] ?? null,
                    'payer_phone' => $ticketData['payer_phone'] ?? null,
                    'vehicle_number' => $ticketData['vehicle_number'] ?? null,
                    'synced' => true,
                    'synced_at' => now(),
                    'updated_at' => now(),
                ]);

            $results[$ticketData['ticket_id']] = ['success' => true];
        }

        return ['success' => true, 'results' => $results];
    }

    protected function generateBatchCode(string $pointCode): string
    {
        $date = date('ymd');
        $random = strtoupper(Str::random(4));
        return "{$pointCode}-{$date}-{$random}";
    }

    protected function generateQRCode(string $ticketNumber): string
    {
        return base64_encode(hash('sha256', $ticketNumber . config('app.key')));
    }

    protected function getSetting(string $key, $default = null)
    {
        $setting = DB::connection('tenant')
            ->table('tenant_settings')
            ->where('key', $key)
            ->first();

        return $setting?->value ?? $default;
    }

    protected function logAudit(string $action, string $module, string $entityType, int $entityId, array $details): void
    {
        DB::connection('tenant')->table('tenant_audit_logs')->insert([
            'user_id' => auth()->id(),
            'user_type' => 'tenant_user',
            'action' => $action,
            'module' => $module,
            'entity_type' => $entityType,
            'entity_id' => $entityId,
            'details' => json_encode($details),
            'ip_address' => request()->ip(),
            'user_agent' => request()->userAgent(),
            'created_at' => now(),
            'updated_at' => now(),
        ]);
    }
}

File: backend/app/services/ticket_service.py
"""
Ticket Service - Manages ticket batches, sales, and verification
"""
from typing import Optional, List
from datetime import datetime, timezone, timedelta
from app.database import db_manager
from app.models.tenant import (
    TicketBatch, TicketBatchCreate, TicketBatchStatus,
    Ticket, TicketStatus, TicketSellRequest, TicketPayment
)
import uuid

class TicketService:
    def __init__(self, tenant_slug: str):
        self.tenant_slug = tenant_slug
        self.db = db_manager.get_tenant_db(tenant_slug)
    
    async def generate_batch_number(self) -> str:
        """Generate unique batch number"""
        year = datetime.now().year
        month = datetime.now().month
        count = await self.db.ticket_batches.count_documents({
            "batch_number": {"$regex": f"^BATCH-{year}{str(month).zfill(2)}-"}
        })
        return f"BATCH-{year}{str(month).zfill(2)}-{str(count + 1).zfill(4)}"
    
    async def create_batch(self, data: TicketBatchCreate, created_by: str) -> TicketBatch:
        """Create a new ticket batch and generate tickets"""
        batch_number = await self.generate_batch_number()
        total_tickets = data.end_number - data.start_number + 1
        total_value = total_tickets * data.unit_price
        
        status = TicketBatchStatus.CREATED
        if data.assigned_to:
            status = TicketBatchStatus.ASSIGNED
        
        batch = TicketBatch(
            id=str(uuid.uuid4()),
            batch_number=batch_number,
            revenue_point_id=data.revenue_point_id,
            revenue_item_id=data.revenue_item_id,
            start_number=data.start_number,
            end_number=data.end_number,
            unit_price=data.unit_price,
            validity_days=data.validity_days,
            status=status,
            assigned_to=data.assigned_to,
            total_tickets=total_tickets,
            tickets_sold=0,
            tickets_cancelled=0,
            total_value=total_value,
            amount_collected=0,
            created_at=datetime.now(timezone.utc),
            created_by=created_by,
        )
        
        batch_dict = batch.model_dump()
        batch_dict['created_at'] = batch_dict['created_at'].isoformat()
        await self.db.ticket_batches.insert_one(batch_dict)
        
        # Generate individual tickets
        tickets_to_insert = []
        for num in range(data.start_number, data.end_number + 1):
            ticket_number = f"{batch_number}-{str(num).zfill(6)}"
            ticket = Ticket(
                id=str(uuid.uuid4()),
                ticket_number=ticket_number,
                batch_id=batch.id,
                revenue_point_id=data.revenue_point_id,
                revenue_item_id=data.revenue_item_id,
                amount=data.unit_price,
                status=TicketStatus.AVAILABLE,
                created_at=datetime.now(timezone.utc),
            )
            ticket_dict = ticket.model_dump()
            ticket_dict['created_at'] = ticket_dict['created_at'].isoformat()
            tickets_to_insert.append(ticket_dict)
        
        if tickets_to_insert:
            await self.db.tickets.insert_many(tickets_to_insert)
        
        return batch
    
    async def assign_batch(self, batch_id: str, collector_id: str) -> Optional[TicketBatch]:
        """Assign a batch to a collector"""
        await self.db.ticket_batches.update_one(
            {"id": batch_id, "status": TicketBatchStatus.CREATED},
            {"$set": {
                "assigned_to": collector_id,
                "status": TicketBatchStatus.ASSIGNED
            }}
        )
        return await self.get_batch(batch_id)
    
    async def sell_ticket(self, data: TicketSellRequest, sold_by: str) -> Optional[Ticket]:
        """Sell a ticket"""
        ticket = await self.get_ticket(data.ticket_id)
        if not ticket or ticket.status != TicketStatus.AVAILABLE:
            return None
        
        # Get batch to check assignment and validity
        batch = await self.get_batch(ticket.batch_id)
        if not batch:
            return None
        
        # Verify seller is assigned to this batch
        if batch.assigned_to and batch.assigned_to != sold_by:
            raise ValueError("You are not assigned to this ticket batch")
        
        # Calculate validity
        valid_until = datetime.now(timezone.utc) + timedelta(days=batch.validity_days)
        
        # Update ticket
        await self.db.tickets.update_one(
            {"id": data.ticket_id},
            {"$set": {
                "status": TicketStatus.SOLD,
                "sold_at": datetime.now(timezone.utc).isoformat(),
                "sold_by": sold_by,
                "valid_until": valid_until.isoformat(),
            }}
        )
        
        # Create payment record
        payment = TicketPayment(
            id=str(uuid.uuid4()),
            ticket_id=data.ticket_id,
            amount=ticket.amount,
            payment_method=data.payment_method,
            payer_name=data.payer_name,
            payer_phone=data.payer_phone,
            collected_by=sold_by,
            created_at=datetime.now(timezone.utc),
        )
        payment_dict = payment.model_dump()
        payment_dict['created_at'] = payment_dict['created_at'].isoformat()
        await self.db.ticket_payments.insert_one(payment_dict)
        
        # Update batch counters
        await self.db.ticket_batches.update_one(
            {"id": ticket.batch_id},
            {"$inc": {"tickets_sold": 1, "amount_collected": ticket.amount}}
        )
        
        # Check if batch should be marked as in_use
        if batch.status == TicketBatchStatus.ASSIGNED:
            await self.db.ticket_batches.update_one(
                {"id": ticket.batch_id},
                {"$set": {"status": TicketBatchStatus.IN_USE}}
            )
        
        return await self.get_ticket(data.ticket_id)
    
    async def cancel_ticket(self, ticket_id: str, reason: str, cancelled_by: str) -> Optional[Ticket]:
        """Cancel a ticket"""
        ticket = await self.get_ticket(ticket_id)
        if not ticket or ticket.status != TicketStatus.AVAILABLE:
            return None
        
        await self.db.tickets.update_one(
            {"id": ticket_id},
            {"$set": {
                "status": TicketStatus.CANCELLED,
                "cancelled_at": datetime.now(timezone.utc).isoformat(),
                "cancelled_by": cancelled_by,
                "cancellation_reason": reason,
            }}
        )
        
        # Update batch counter
        await self.db.ticket_batches.update_one(
            {"id": ticket.batch_id},
            {"$inc": {"tickets_cancelled": 1}}
        )
        
        return await self.get_ticket(ticket_id)
    
    async def verify_ticket(self, ticket_number: str) -> dict:
        """Verify a ticket is valid"""
        ticket = await self.db.tickets.find_one({"ticket_number": ticket_number}, {"_id": 0})
        if not ticket:
            return {"valid": False, "reason": "Ticket not found"}
        
        if ticket['status'] == TicketStatus.AVAILABLE:
            return {"valid": False, "reason": "Ticket not sold"}
        
        if ticket['status'] == TicketStatus.CANCELLED:
            return {"valid": False, "reason": "Ticket has been cancelled"}
        
        if ticket['status'] == TicketStatus.EXPIRED:
            return {"valid": False, "reason": "Ticket has expired"}
        
        # Check validity period
        if ticket.get('valid_until'):
            valid_until = datetime.fromisoformat(ticket['valid_until'])
            if datetime.now(timezone.utc) > valid_until:
                await self.db.tickets.update_one(
                    {"id": ticket['id']},
                    {"$set": {"status": TicketStatus.EXPIRED}}
                )
                return {"valid": False, "reason": "Ticket has expired"}
        
        return {
            "valid": True,
            "ticket_number": ticket['ticket_number'],
            "amount": ticket['amount'],
            "sold_at": ticket.get('sold_at'),
            "valid_until": ticket.get('valid_until'),
        }
    
    async def get_batch(self, batch_id: str) -> Optional[TicketBatch]:
        """Get batch by ID"""
        doc = await self.db.ticket_batches.find_one({"id": batch_id}, {"_id": 0})
        if doc:
            if isinstance(doc.get('created_at'), str):
                doc['created_at'] = datetime.fromisoformat(doc['created_at'])
            return TicketBatch(**doc)
        return None
    
    async def get_ticket(self, ticket_id: str) -> Optional[Ticket]:
        """Get ticket by ID"""
        doc = await self.db.tickets.find_one({"id": ticket_id}, {"_id": 0})
        if doc:
            if isinstance(doc.get('created_at'), str):
                doc['created_at'] = datetime.fromisoformat(doc['created_at'])
            return Ticket(**doc)
        return None
    
    async def list_batches(self, skip: int = 0, limit: int = 100,
                          status: TicketBatchStatus = None,
                          assigned_to: str = None) -> List[TicketBatch]:
        """List ticket batches"""
        query = {}
        if status:
            query["status"] = status
        if assigned_to:
            query["assigned_to"] = assigned_to
        
        cursor = self.db.ticket_batches.find(query, {"_id": 0}).skip(skip).limit(limit).sort("created_at", -1)
        batches = []
        async for doc in cursor:
            if isinstance(doc.get('created_at'), str):
                doc['created_at'] = datetime.fromisoformat(doc['created_at'])
            batches.append(TicketBatch(**doc))
        return batches
    
    async def list_tickets(self, batch_id: str = None, status: TicketStatus = None,
                          skip: int = 0, limit: int = 100) -> List[Ticket]:
        """List tickets"""
        query = {}
        if batch_id:
            query["batch_id"] = batch_id
        if status:
            query["status"] = status
        
        cursor = self.db.tickets.find(query, {"_id": 0}).skip(skip).limit(limit)
        tickets = []
        async for doc in cursor:
            if isinstance(doc.get('created_at'), str):
                doc['created_at'] = datetime.fromisoformat(doc['created_at'])
            tickets.append(Ticket(**doc))
        return tickets
    
    async def get_collector_tickets(self, collector_id: str, include_sold: bool = False) -> List[Ticket]:
        """Get tickets assigned to a collector"""
        # Get batches assigned to collector
        batch_ids = []
        async for batch in self.db.ticket_batches.find({"assigned_to": collector_id}, {"id": 1}):
            batch_ids.append(batch['id'])
        
        if not batch_ids:
            return []
        
        query = {"batch_id": {"$in": batch_ids}}
        if not include_sold:
            query["status"] = TicketStatus.AVAILABLE
        
        cursor = self.db.tickets.find(query, {"_id": 0})
        tickets = []
        async for doc in cursor:
            if isinstance(doc.get('created_at'), str):
                doc['created_at'] = datetime.fromisoformat(doc['created_at'])
            tickets.append(Ticket(**doc))
        return tickets
    
    async def get_ticket_stats(self) -> dict:
        """Get ticket statistics"""
        pipeline = [
            {"$group": {
                "_id": "$status",
                "count": {"$sum": 1},
                "total_value": {"$sum": "$amount"}
            }}
        ]
        
        stats = {
            "total_tickets": 0,
            "total_value": 0,
            "by_status": {}
        }
        
        async for result in self.db.tickets.aggregate(pipeline):
            status = result['_id']
            stats['by_status'][status] = {
                "count": result['count'],
                "value": result['total_value']
            }
            stats['total_tickets'] += result['count']
            if status == TicketStatus.SOLD:
                stats['total_value'] += result['total_value']
        
        return stats
